File: MatrixScanBubblesSample/src/main/java/com/scandit/datacapture/matrixscanbubblessample/scan/bubble/Bubble.java
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.scandit.datacapture.matrixscanbubblessample.scan.bubble;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.transition.Fade;
import androidx.transition.Transition;
import androidx.transition.TransitionManager;
import com.scandit.datacapture.matrixscanbubblessample.R;
import com.scandit.datacapture.matrixscanbubblessample.scan.bubble.data.BubbleData;

import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;

public class Bubble {

    public final View root;
    private final View containerShelfData;
    private final TextView textCode;

    private Transition transition = new Fade();
    private boolean showShelfData = true;

    public Bubble(Context context, BubbleData data, String code) {
        root = View.inflate(context, R.layout.bubble, null);
        root.setLayoutParams(new ViewGroup.LayoutParams(WRAP_CONTENT, WRAP_CONTENT));
        containerShelfData = root.findViewById(R.id.shelf_container);
        textCode = root.findViewById(R.id.text_code);
        TextView textShelfData = root.findViewById(R.id.text_shelf_data);

        transition.setDuration(100);

        // We want to show the scanned code when tapping on the bubble. To do so, it's enough to
        // just add a regular View.OnClickListener on the view.
        root.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                toggleText();
            }
        });

        textShelfData.setText(
                context.getString(R.string.shelf_text, data.shelfCount, data.backroomCount)
        );
        textCode.setText(code);

        showCurrentData();
    }

    private void startTransition() {
        TransitionManager.beginDelayedTransition((ViewGroup) root, transition);
    }

    private void toggleText() {
        showShelfData = !showShelfData;
        showCurrentData();
    }

    private void showCurrentData() {
        if (showShelfData) {
            showShelfDataInternal();
        } else {
            showCodeInternal();
        }
    }

    public void show() {
        startTransition();
        root.setVisibility(View.VISIBLE);
    }

    public void hide() {
        startTransition();
        root.setVisibility(View.GONE);
    }

    private void showShelfDataInternal() {
        containerShelfData.setVisibility(View.VISIBLE);
        textCode.setVisibility(View.GONE);
    }

    private void showCodeInternal() {
        containerShelfData.setVisibility(View.GONE);
        textCode.setVisibility(View.VISIBLE);
    }
}

File: InventoryAuditSample/src/main/java/com/scandit/datacapture/inventoryauditsample/scan/bubble/Bubble.kt
/*
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.scandit.datacapture.inventoryauditsample.scan.bubble

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.widget.EditText
import androidx.core.content.ContextCompat.getDrawable
import com.scandit.datacapture.inventoryauditsample.R
import com.scandit.datacapture.inventoryauditsample.scan.bubble.data.BubbleColor.RED
import com.scandit.datacapture.inventoryauditsample.scan.bubble.data.BubbleColor.WHITE
import com.scandit.datacapture.inventoryauditsample.scan.bubble.data.BubbleData
import com.scandit.datacapture.inventoryauditsample.scan.bubble.data.BubbleDataProvider

class Bubble(context: Context, data: BubbleData, private val code: String) : TextWatcher {
    private val backgrounds = mapOf(
            RED to getDrawable(context, R.drawable.bubble_background_red),
            WHITE to getDrawable(context, R.drawable.bubble_background_white)
    )

    val root: View = (View.inflate(context, R.layout.bubble, null) as EditText).apply {
        layoutParams = ViewGroup.LayoutParams(WRAP_CONTENT, WRAP_CONTENT)
        background = backgrounds.getValue(data.color)
        setText(data.count.toString())
        addTextChangedListener(this@Bubble)
    }

    override fun afterTextChanged(s: Editable?) {
        if (s.toString().isNotBlank()) {
            root.background = backgrounds.getValue(WHITE)
            BubbleDataProvider.saveDataForBarcode(code, s.toString().toInt())
        }
    }

    override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
        // Not interested in this callback.
    }

    override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
        // Not interested in this callback.
    }
}